src/units.rs

```diff
@@ -1,116 +1,109 @@
+    pub(crate) fn new(parent_map: &NodeMap<Option<H::Hash>>) -> Self {
+        let hash = Self::combine_hashes(&parent_map);
+        let parents = parent_map.iter().map(Option::is_some).collect();
+
+        ControlHash {
+            parents_mask: parents,
+            hash,
+        }
     }
-}
 
-pub(crate) type UncheckedSignedUnit<H, D, S> = UncheckedSigned<FullUnit<H, D>, S>;
+    pub(crate) fn combine_hashes(parent_map: &NodeMap<Option<H::Hash>>) -> H::Hash {
+        parent_map.using_encoded(H::hash)
+    }
 
-pub(crate) type SignedUnit<'a, H, D, KB> = Signed<'a, FullUnit<H, D>, KB>;
+    pub(crate) fn parents(&self) -> impl Iterator<Item = NodeIndex> + '_ {
+        self.parents_mask
+            .iter()
+            .enumerate()
+            .filter(|(_, b)| *b)
+            .map(|(i, _)| i.into())
+    }
 
-#[derive(Debug, Clone, Copy, Eq, PartialEq, Encode, Decode, StdHash)]
-pub(crate) struct UnitCoord {
-    pub(crate) creator: NodeIndex,
-    pub(crate) round: u64,
-}
+    pub(crate) fn n_parents(&self) -> NodeCount {
+        NodeCount(self.parents().count())
+    }
 
-impl<H: Hasher> From<Unit<H>> for UnitCoord {
-    fn from(unit: Unit<H>) -> Self {
-        UnitCoord {
-            creator: unit.creator(),
-            round: unit.round() as u64,
-        }
+    pub(crate) fn n_members(&self) -> NodeCount {
+        NodeCount(self.parents_mask.len())
     }
 }
 
-impl<H: Hasher> From<&Unit<H>> for UnitCoord {
-    fn from(unit: &Unit<H>) -> Self {
-        UnitCoord {
-            creator: unit.creator(),
-            round: unit.round() as u64,
-        }
+impl<H: Hasher> Encode for ControlHash<H> {
+    fn encode_to<T: Output + ?Sized>(&self, dest: &mut T) {
+        (self.parents_mask.len() as u32).encode_to(dest);
+        self.parents_mask.to_bytes().encode_to(dest);
+        self.hash.encode_to(dest);
     }
 }
 
-impl From<(usize, NodeIndex)> for UnitCoord {
-    fn from(coord: (usize, NodeIndex)) -> Self {
-        UnitCoord {
-            creator: coord.1,
-            round: coord.0 as u64,
-        }
+impl<H: Hasher> Decode for ControlHash<H> {
+    fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+        let len = u32::decode(input)?;
+        let bytes = Vec::decode(input)?;
+        let mut parents = bit_vec::BitVec::from_bytes(&bytes);
+        parents.truncate(len as usize);
+        let hash = H::Hash::decode(input)?;
+        Ok(ControlHash {
+            parents_mask: parents,
+            hash,
+        })
     }
 }
 
-type UnitRound = u64;
-
-#[derive(Clone, Debug, PartialEq, Encode, Decode)]
+/// The simplest type representing a unit, consisting
+#[derive(Clone, Debug, Default, PartialEq, Encode, Decode)]
 pub(crate) struct PreUnit<H: Hasher> {
-    pub(crate) creator: NodeIndex,
-    pub(crate) round: UnitRound,
-    pub(crate) control_hash: ControlHash<H>,
+    coord: UnitCoord,
+    // round is stored as `u16` not `Round = usize` for efficiency. Also, `usize` does not implement
```

Comment:
Maybe the conversion should happen in encoding? Not sure because most probably compiler is smart enough to optimize this anyway.

Reply:
I would rather avoid implementing `Encode + Decode` manually. There is an attribute `#[codec(encoded_as = "OtherType")]`, but it requres implementing `EncodeAsRef` trait, and that would require introducing wrappers (implementing `EncodeAsRef<'_, usize>` for `u16` is just wrong).